### PR TITLE
[NEW] Add support for custom emojis in reactions

### LINKED ISSUE #699: [NEW] Add support to Custom Emojis
There's a web socket call that returns all custom emojis from the server, you can read more about it here: https://rocket.chat/docs/developer-guides/realtime-api/method-calls/list-custom-emoji/#list-custom-emoji.

### PR DESCRIPTION
**Pull request by cardoso**
@RocketChat/ios

Preview:
https://imgur.com/a/zwjYM

Closes #699

### CODE REVIEW

File: Rocket.Chat/Controllers/Chat/ChatControllerMessageActions.swift [hunk 1]
```diff
@@ -24,15 +25,24 @@ extension ChatViewController {
                 presenter.sourceRect = view.bounds
             }
 
+            controller.emojiPicked = { emoji in
+                MessageManager.react(message, emoji: emoji, completion: { _ in })
+            }
+
+            controller.customEmojis = {
+                guard let emojis = Realm.shared?.objects(CustomEmoji.self) else { return [] }
```
Comment (rafaelks) [inline Rocket.Chat/Controllers/Chat/ChatControllerMessageActions.swift]: We could avoid importing `RealmSwift` here by just having a method in `CustomEmoji` that returns the objects as Emojis. What do you think?

File: Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift [hunk 2]
```diff
@@ -7,6 +7,7 @@
 //
 
 import UIKit
+import RealmSwift
```
Comment (rafaelks) [inline Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift]: I believe you missed this import 🤔

File: Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift [hunk 3]
```diff
@@ -271,8 +272,13 @@ final class ChatMessageCell: UICollectionViewCell {
                 highlight = false
             }
 
+            let emoji = reaction.emoji ?? "?"
+
+            let imageUrl = CustomEmoji.withShortname(emoji)?.imageUrl()
```
Comment (rafaelks) [inline Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift]: Could remove this empty line?

File: Rocket.Chat/Managers/Model/CustomEmojiManager.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            "method": "listEmojiCustom",
+            "params": []
+            ] as [String: Any]
+
+        SocketManager.send(requestEmojis) { response in
+            guard !response.isError() else { return Log.debug(response.result.string) }
+
+            let emojis = List<CustomEmoji>()
+
+            let list = response.result["result"].array
```
Comment (rafaelks) [inline Rocket.Chat/Managers/Model/CustomEmojiManager.swift]: Could remove this empty line?

File: Rocket.Chat/Views/EmojiPicker/EmojiPicker.swift [hunk 5]
```diff
@@ -61,27 +77,13 @@ class EmojiPicker: UIView {
     }
 
     fileprivate var currentCategories: [EmojiCategory] {
-        let recentCategory = (name: "recent", emojis: recentEmojis)
-        let categories = (recentEmojis.count > 0 ? [recentCategory] : []) + defaultCategories
+        let categories = (recentEmojis.count > 0 ? [recentCategory] : []) + (customEmojis.count > 0 ? [customCategory] : []) + defaultCategories
```
Comment (rafaelks) [inline Rocket.Chat/Views/EmojiPicker/EmojiPicker.swift]: What about doing all this in three steps?

File: Rocket.Chat/Views/EmojiPicker/EmojiPicker.swift [hunk 6]
```diff
@@ -33,6 +39,16 @@ class EmojiPicker: UIView {
         }
     }
 
+    var recentCategory: (name: String, emojis: [Emoji]) {
+        // remove invalid custom emojis
+        let recentEmojis = self.recentEmojis.filter {
+            guard let imageUrl = $0.imageUrl else { return true }
```
Comment (rafaelks) [inline Rocket.Chat/Views/EmojiPicker/EmojiPicker.swift]: Wouldn't be `return false` here?

**Comment by rafaelks** [comment]
Added a few comments to you.

File: Rocket.Chat/Views/EmojiPicker/EmojiPicker.swift (hunk 6 above)
Comment (cardoso) [reply Rocket.Chat/Views/EmojiPicker/EmojiPicker.swift]: If imageUrl is not nil, means it's not a custom emoji and we don't need to check if it's invalid, so we return true for the filter.

Comment (rafaelks) [reply Rocket.Chat/Views/EmojiPicker/EmojiPicker.swift]: Gotcha! What do you think about creating a property/method/getter called `isCustomEmoji` and then use it? I think it'll be more clear to understand later 😂

**Comment by cardoso** [comment]
Nice suggestions @rafaelks! Applied them.

**Comment by rafaelks** [comment]
@cardoso Thanks, let me know when it's ready to review again please.

**Comment by cardoso** [comment]
It's ready @rafaelks 😛

**Pull request merged**